### PR TITLE
Snap Deals full unseal

### PR DESCRIPTION
**Pull request by geoff-vball**
## Related Issues
resolves https://github.com/filecoin-project/lotus/issues/8230

## Proposed Changes
Lotus can now recover unsealed data from a replica-update sector if the sector key has been deleted.

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [ ] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

File: extern/sector-storage/manager.go
Comment (ZenGround0) [inline extern/sector-storage/manager.go:286]: I'm still confused about the importance of doing worker Fetch vs delegating to the sealer_cgo sealer to access sector files, but something looks off here.  In the case you have both the sector key and the update replica (true for the first finality of a sectors lifetime + for all subsequent unseals after the first one) the first call will pass and the second fetch won't happen.  

I think this ends up being fine because the sealer will fetch anyways later but then it makes me wonder why we put anything in here at all.

File: extern/sector-storage/ffiwrapper/sealer_cgo.go
Comment (ZenGround0) [inline extern/sector-storage/ffiwrapper/sealer_cgo.go:273]: You don't want to give it this name because it implies we're dropping the sector key file (as the Sealer function of the same name does).  

We probably *do* want to actually drop the sector key (after this we will already be keeping around unsealed) by implementing Sealer.ReleaseSectorKey and then calling it here.  @magik6k @Reiers any objections?

**Review by ZenGround0 — COMMENTED** [review]
A few things but this is close to ready

File: extern/sector-storage/ffiwrapper/sealer_cgo.go
Comment (geoff-vball) [reply extern/sector-storage/ffiwrapper/sealer_cgo.go:308]: Issue linked [here](https://github.com/filecoin-project/lotus/issues/8487). The crux is there is a file called `p_aux` needed by `DecodeFrom` that is generated in `SealPreCommit2`

File: extern/sector-storage/manager.go
Comment (geoff-vball) [reply extern/sector-storage/manager.go:286]: Let's talk about this one sometime tomorrow. I think it's complicated because we're trying to share the same code path for unsealing replica updates with normal sealed sectors and it's making things messy.

**Review by ZenGround0 — APPROVED** [review]
LGTM after a bit of out of band analysis

File: extern/sector-storage/ffiwrapper/sealer_cgo.go
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/sealer_cgo.go:308]: Wait, the original cache with p_aux should exist even after releasing sector key

Comment (magik6k) [reply extern/sector-storage/ffiwrapper/sealer_cgo.go:308]: (or, do we remove it in FinalizeSector?)

Comment (geoff-vball) [reply extern/sector-storage/ffiwrapper/sealer_cgo.go:308]: Okay, maybe we just have to make a call to `AcquireSector` to potentially fetch the cache from a remote to get p_aux. I'll look at this today.

Comment (geoff-vball) [reply extern/sector-storage/ffiwrapper/sealer_cgo.go:308]: Okay, we do keep the cache in `FinalizeSector`, it's currently getting removed with our call to `SealPreCommit1`, which removes anything in the cache folder. @magik6k do you think we should create a modified `SealPreCommit1` that doesn't do this?

File: lib/rpcenc/reader.go
Comment (magik6k) [inline lib/rpcenc/reader.go:18]:
```suggestion
	"github.com/filecoin-project/lotus/extern/storage-sealing/lib/nullreader"
```
+rename below

File: lib/rpcenc/reader_test.go
Comment (magik6k) [inline lib/rpcenc/reader_test.go:12]:
```suggestion
	"github.com/filecoin-project/lotus/extern/storage-sealing/lib/nullreader"
```

File: extern/sector-storage/manager.go
Comment (magik6k) [inline extern/sector-storage/manager.go:285]: Those errors won't get propagated, would need https://github.com/filecoin-project/go-jsonrpc/pull/36 (Maybe worth finally landing for this?)

File: extern/sector-storage/ffiwrapper/sealer_cgo.go
Comment (magik6k) [inline extern/sector-storage/ffiwrapper/sealer_cgo.go:18]: Nit: We're generally trying to keep imports ordered in groups like this:
```
std

libs

filecoin-project/not-lotus

filecoin-project/lotus
```

So this should be at the bottom (..ideally we'd convince some linter to do this for us)

**Review by magik6k — COMMENTED** [review]
Generally looks good, just a few comments.

Looks like there is unintentional `extern/filecoin-ffi` change committed.

Snapdeals test appears to be unhappy:
```
2022-04-15T14:58:54.849Z	WARN	stores	stores/local.go:599	RemoveCopies: no primary copies of sector {1000 1} (sealed), not removing anything
2022-04-15T14:58:54.849Z	WARN	stores	stores/remote.go:73	remote RemoveCopies: no primary copies of sector {1000 1} (sealed), not removing anything
2022-04-15T14:58:54.849Z	WARN	stores	stores/local.go:599	RemoveCopies: no primary copies of sector {1000 1} (cache), not removing anything
2022-04-15T14:58:54.849Z	WARN	stores	stores/remote.go:73	remote RemoveCopies: no primary copies of sector {1000 1} (cache), not removing anything
Release Sector Key
Unseal Replica
    manager_test.go:323: 
        	Error Trace:	manager_test.go:323
        	Error:      	Received unexpected error:
        	            	worker UnsealPiece call:
        	            	    github.com/filecoin-project/lotus/extern/sector-storage.(*Manager).SectorsUnsealPiece
        	            	        /home/circleci/project/extern/sector-storage/manager.go:327
        	            	  - storage call error 0: %!w(failed to acquire sector {1000 1} from remote(2): sector not found [Hostname: cba2a59638c7])
        	Test:       	TestSnapDeals
--- FAIL: TestSnapDeals (24.09s)
```

File: extern/sector-storage/ffiwrapper/sealer_cgo.go
Comment (ZenGround0) [reply extern/sector-storage/ffiwrapper/sealer_cgo.go:18]: check out go-fil-markets for at least one approach, I noticed they had automatic dependency rearranging when making a PR there last week

**Comment by jennijuju** [comment]
@geoff-vball could you plz update the pr title format as suggested in the template?

**Comment by arajasek** [comment]
Blocked by https://github.com/filecoin-project/go-jsonrpc/pull/36

File: extern/sector-storage/manager.go
Comment (geoff-vball) [reply extern/sector-storage/manager.go:285]: @magik6k I found the issue that was causing these errors to be returned as strings and it is fixed in the latest commit. Is there still a codepath where this might hit RPC? If so, could you show me where it is?

Comment (magik6k) [reply extern/sector-storage/manager.go:285]: worker.Fetch is an rpc call.

This will only work with the miner built-in worker, as that one doesn't go through rpc.

Comment (geoff-vball) [reply extern/sector-storage/manager.go]: @magik6k Talked this over with Zen and we decided that not checking the error type will still give us the desired result in all cases. If you agree, I think we can land this PR

**Review by magik6k — APPROVED** [review]
I don't see anything obviously wrong here, and it's definitely better that the current state of things.

**Pull request merged**
